pkg/binary/envoy/debug/lsof.go

```diff
@@ -0,0 +1,130 @@
+	User    string `json:"user"`
+	Fd      string `json:"fd"`
+	// non process dependent information
+	Type   string `json:"type"`   // type of node associated with the file / FIFO for FIFO special file
+	Device string `json:"device"` // device numbers associated with the file, separated by commas
+	Size   string `json:"size"`
+	Node   string `json:"node"` // inode number of a local file/ Internet protocol type
+	Name   string `json:"name"` // name of the mount point and file system on which the file resides
+}
+
+// EnableOpenFilesDataCollection is a preset option that registers collection of statistics of files opened by envoy instance(s)
+func EnableOpenFilesDataCollection(r *envoy.Runtime) {
+	if err := os.Mkdir(filepath.Join(r.DebugStore(), "lsof"), os.ModePerm); err != nil {
+		log.Errorf("error in creating a directory to write open file data of envoy to: %v", err)
+	}
+	r.RegisterPreTermination(retrieveOpenFilesData)
+}
+
+// retrieveOpenFilesData writes statistics of open files associated with envoy instance(s) to a json file
+// if succeeded, return nil, else return an error instance
+func retrieveOpenFilesData(r binary.Runner) error { //nolint:gocyclo
+	// get a list of processes
+	processes, err := process.Processes()
+	if err != nil {
+		return fmt.Errorf("error in getting process pids")
+	}
+
+	// filter Process instances of envoy
+	isEnvoy := func(p *process.Process) bool {
+		name, err := p.Name() //nolint:govet
+		if err != nil {
+			log.Errorf("error in getting process name for %v", p)
+			return false
+		}
+		return name == "envoy"
+	}
+	envoys := make([]*process.Process, 0)
+	for _, p := range processes {
+		if isEnvoy(p) {
+			envoys = append(envoys, p)
+		}
+	}
+
+	f, err := os.Create(filepath.Join(r.DebugStore(), "lsof/lsof.json"))
+	if err != nil {
+		return fmt.Errorf("unable to create file to write lisof output to: %v", err)
+	}
+	defer f.Close() //nolint
+
+	ofStatArr := make([]Lsof, 0)
+	// print open file stats for all envoy instances
+	for _, envoy := range envoys {
+		// relevant fields of the process
+		username, _ := envoy.Username()
+		name, _ := envoy.Name()
+		pid := envoy.Pid
+
+		openFiles, err := envoy.OpenFiles() //nolint:govet
+		if err != nil {
+			log.Debugf("error in getting ofStat for %v\n", envoy)
+			continue
+		}
+
+		for _, stat := range openFiles {
+			ofStat := Lsof{
+				Command: name,
+				Pid:     fmt.Sprint(pid),
+				User:    username,
+				Fd:      fmt.Sprint(stat.Fd),
+				Name:    stat.Path,
+			}
+
+			var fstat syscall.Stat_t
+			if err := syscall.Stat(stat.Path, &fstat); err != nil {
+				// continue if the path is invalid
+				ofStatArr = append(ofStatArr, ofStat)
+				continue
+			}
+
+			ofStat.Node = fmt.Sprint(fstat.Ino)
+			ofStat.Size = fmt.Sprint(fstat.Size)
+			ofStatArr = append(ofStatArr, ofStat)
+		}
+	}
+
+	out, err := json.Marshal(ofStatArr)
+	if err != nil {
+		return fmt.Errorf("unable to convert to json representation: %v", err)
+	}
+	// write to file
```

Comment:
Comments should generally only explain why not what, Go is verbose by design to avoid what comments.